### PR TITLE
ddl, util/types: make converter correctly convert string to hex or bit.

### LINKED ISSUE #2717: Failed to `insert t1 values('\0');`
1. What did you do?
create table t1 (a bit(1));
insert t1 values('\0');

2. What did you expect to see?
Query OK, 1 row affected (0.01 sec)

3. What did you see instead?
ERROR 1265 (01000): Data Truncated

### LINKED ISSUE #2858: Failed to `create table t (a bit(1) default b'0')`
1. What did you do?
create table t (a bit(1) default b'0');

2. What did you expect to see?
Query OK, 0 rows affected (0.01 sec)

3. What did you see instead?
ERROR 1067 (42000): Invalid default value for 'a'

### PR DESCRIPTION
**Pull request by bobotu**
This PR fix #2858 and #2717. Furthermore fix the error that
unable to set default value as a hex value just like what happend in #2858.
I also noticed that TiDB have some differences when compared to MySQL.

create table t1 (a bit(64));
insert t1 value ('10');
select a+0 from t1;

In MySQL we can get:
+-------+
| a+0   |
+-------+
| 12592 |
+-------+

But in TiDB we will get:
+-------+
| a+0   |
+-------+
| 10    |
+-------+

As a result of fix #2717, TiDB now behave same as MySQL.

### CODE REVIEW

**Comment by ngaut** [comment]
Well done.

**Comment by zhexuany** [comment]
@bobotu Please rebase your PR and do run `goword` to ensure there is no comments or spelling error.

File: util/types/datum.go [hunk 1]
```diff
@@ -1044,7 +1044,17 @@ func (d *Datum) convertToMysqlYear(sc *variable.StatementContext, target *FieldT
 }
 
 func (d *Datum) convertToMysqlBit(sc *variable.StatementContext, target *FieldType) (Datum, error) {
-	x, err := d.convertToUint(sc, target)
+	var (
```
Comment (shenli) [inline util/types/datum.go]: Any unit test covers this?

Comment (bobotu) [inline util/types/datum.go]: Newly added tests are fundamentally related to changes within this function. Parser will pass a string type `Datum`  to this function when user input `'\0'` or `'100'`, and `getDefaultValue` will pass a string presentation of `bit` to it in the end.  As mentioned in commit message, MySQL treat string as binary string when convert it to `bit`, but TiDB treat it as number literal. This difference cause #2858 and #2717. So new tests cover the string type case and existing tests cover others cases.

**Comment by coocood** [comment]
LGTM
Great job!

File: ddl/ddl_db_test.go [hunk 2]
```diff
@@ -1093,3 +1093,21 @@ out:
...
 }
+
+func (s *testDBSuite) TestIssue2858And2717(c *C) {
+	defer testleak.AfterTest(c)()
+	s.tk = testkit.NewTestKit(c, s.store)
+	s.tk.MustExec("use " + s.schemaName)
+
+	s.tk.MustExec("create table t_issue_2858_bit (a bit(64) default b'0')")
+	s.tk.MustExec("insert into t_issue_2858_bit value ()")
+	s.tk.MustExec(`insert into t_issue_2858_bit values (100), ('10'), ('\0')`)
```
Comment (XuHuaiyu) [inline ddl/ddl_db_test.go]: add test for 
insert into t_issue_2858_bit  value('\x01')
select a + 0 from t;

In MySQL, we get: 7876657

It seems we will get 2 in TiDB with this commmit?

File: util/types/bit.go [hunk 3]
```diff
@@ -104,3 +105,33 @@ func ParseBit(s string, width int) (Bit, error) {
 
 	return Bit{Value: n, Width: width}, nil
 }
+
+// ParseBitValue parses the binary string for bit type into uint64.
+func ParseBitValue(s string, width int) (uint64, error) {
```
Comment (XuHuaiyu) [inline util/types/bit.go]: We already have a func named ParseBit to parse binary string to uint64.
It seems the implementation of this func is used to parse a hex string to uint64?
If so, the width in line 117 may be width = len(b) * 8?

File: util/types/bit.go [hunk 4]
```diff
@@ -104,3 +105,33 @@ func ParseBit(s string, width int) (Bit, error) {
...
+	if len(s) == 0 {
+		return 0, errors.Errorf("invalid empty string for parsing bit value")
+	}
+
+	b := hack.Slice(s)
+	if width == UnspecifiedBitWidth {
+		width = (len(b) + 7) & ^7
```
Comment (XuHuaiyu) [inline util/types/bit.go]: The implementation of this func seems to be parse `hex` string to uint64?
If so, it seems width = len(b)*8?

Comment (bobotu) [inline util/types/bit.go]: You are right. I'll fix it later.

File: ddl/ddl_db_test.go (hunk 2 above)
Comment (XuHuaiyu) [inline ddl/ddl_db_test.go]: Never mind it, I've test this case which is ok.

File: util/types/bit.go (hunk 4 above)
Comment (XuHuaiyu) [inline util/types/bit.go]: ok, would it be better to rename the function to `ParseHex`?

Comment (XuHuaiyu) [inline util/types/bit.go]: or ParseStringToBitValue

Comment (bobotu) [inline util/types/bit.go]: `ParseStringToBitValue` would be better. I'll refactor the code.

File: util/types/bit.go [hunk 5]
```diff
@@ -106,15 +106,15 @@ func ParseBit(s string, width int) (Bit, error) {
 	return Bit{Value: n, Width: width}, nil
 }
 
-// ParseBitValue parses the binary string for bit type into uint64.
-func ParseBitValue(s string, width int) (uint64, error) {
+// ParseStringToBitValue parses the binary string for bit type into uint64.
```
Comment (XuHuaiyu) [inline util/types/bit.go]: s/binary string/string

Comment (bobotu) [inline util/types/bit.go]: Sorry I don‘t understand what you mean.

Comment (tiancaiamao) [inline util/types/bit.go]: substitute "binary string" with "string" usually written as "s/binary string/string" for short.

**Comment by bobotu** [comment]
PTAL @shenli

**Comment by shenli** [comment]
LGTM
@bobotu Thanks! Good Job!

**Pull request merged**